components/tidb_query/src/codec/collation/mod.rs

```diff
@@ -91,7 +93,128 @@ impl Collator for CollatorBinary {
     }
 
     #[inline]
-    fn sort_hash<H: Hasher>(_state: &mut H, _bstr: &[u8]) -> Result<()> {
-        unimplemented!()
+    fn sort_hash<H: Hasher>(state: &mut H, bstr: &[u8]) -> Result<()> {
+        bstr.hash(state);
+        Ok(())
+    }
+}
+
+#[derive(Debug)]
+#[repr(transparent)]
+pub struct SortKey<T, C: Collator>
+where
+    T: AsRef<[u8]>,
+{
+    inner: T,
+    _phantom: PhantomData<C>,
+}
+
+impl<T, C: Collator> SortKey<T, C>
+where
+    T: AsRef<[u8]>,
+{
+    #[inline]
+    pub fn new(inner: T) -> Result<Self> {
+        C::validate(inner.as_ref())?;
+        Ok(Self {
+            inner,
+            _phantom: PhantomData,
+        })
+    }
+
+    /// Create SortKey from unchecked bytes.
+    ///
+    /// # Safety
+    ///
+    /// The `Ord`, `Hash`, `PartialEq` and more implementations assume that the bytes are
+    /// valid for the certain collator. The violation will cause panic.
+    #[inline]
+    pub unsafe fn new_unchecked(inner: T) -> Self {
+        Self {
+            inner,
+            _phantom: PhantomData,
+        }
+    }
+
+    #[inline]
+    #[allow(clippy::transmute_ptr_to_ptr)]
+    pub fn map_option(inner: &Option<T>) -> Result<&Option<Self>> {
+        if let Some(inner) = inner {
+            C::validate(inner.as_ref())?;
+        }
+        Ok(unsafe { std::mem::transmute(inner) })
+    }
+
+    #[inline]
+    pub fn into_inner(self) -> T {
+        self.inner
+    }
+}
+
+impl<T, C: Collator> Hash for SortKey<T, C>
+where
+    T: AsRef<[u8]>,
+{
+    #[inline]
+    fn hash<H: Hasher>(&self, state: &mut H) {
+        C::sort_hash(state, self.inner.as_ref()).unwrap()
+    }
+}
+
+impl<T, C: Collator> PartialEq for SortKey<T, C>
+where
+    T: AsRef<[u8]>,
+{
+    #[inline]
+    fn eq(&self, other: &Self) -> bool {
+        C::sort_compare(&self.inner.as_ref(), &other.inner.as_ref()).unwrap()
+            == std::cmp::Ordering::Equal
```

Comment:
Could be there any possibility that the collator fails while hashing/comparing SortKeys?. If so, unwrapping a result seems will cause a panic.

Reply:
It'll be checked when initialize, unless `new_unchecked` is used.

Reply:
OK! I see.